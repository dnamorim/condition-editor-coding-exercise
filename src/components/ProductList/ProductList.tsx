import {
  Table,
  TableHeader,
  TableRow,
  TableCell,
  TableHeaderCell,
  TableBody,
} from 'semantic-ui-react';
import { PropTypes } from './ProductList.types';

const ProductList = ({ products, properties }: PropTypes) => {
  const maxcolumns = properties.length;

  return (
    <Table striped celled>
      <TableHeader>
        <TableRow>
          {properties.map(({ key, text }) => (
            <TableHeaderCell key={`th_${key}`}>{text}</TableHeaderCell>
          ))}
        </TableRow>
      </TableHeader>

      <TableBody>
        {products.map(({ id, properties: productProperties }) => {
          return (
            <TableRow key={`tr_${id}`}>
              {properties.map(({ key }) => (
                <TableCell key={`tc_prod${id}_${key}`}>
                  {productProperties[key]}
                </TableCell>
              ))}
              {properties.length < maxcolumns ? <TableCell></TableCell> : <></>}
            </TableRow>
          );
        })}
      </TableBody>
    </Table>
  );
};

export default ProductList;
