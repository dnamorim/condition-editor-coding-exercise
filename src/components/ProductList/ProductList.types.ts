import { ProductListItem } from 'data/selectors/types';

export type PropTypes = {
  products: ProductListItem[];
  properties: { [key: string]: any }[];
};
