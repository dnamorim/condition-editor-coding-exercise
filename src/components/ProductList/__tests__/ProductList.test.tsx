import { render, screen } from '@testing-library/react';
import ProductList from '..';

describe('ProductList', () => {
  const mockProps = {
    properties: [
      {
        key: 0,
        text: 'Product Name',
      },
      {
        key: 1,
        text: 'Color',
      },
    ],
    products: [
      {
        id: 0,
        properties: {
          0: 'Foo',
          1: 'Bar',
        },
      },
      {
        id: 1,
        properties: {
          0: 'Bizz',
          1: 'Buzz',
        },
      },
    ],
  };

  it('should render correctly', () => {
    const { container } = render(<ProductList {...mockProps} />);
    const renderedProducts = screen.getAllByRole('rowgroup')[1].childNodes;

    expect(renderedProducts).toHaveLength(mockProps.products.length);
    expect(container).toMatchSnapshot();
  });

  it("should render an empty cell when a property value doesn't exist", () => {
    const extraProduct = {
      id: 2,
      properties: { 0: 'Sample' },
    };
    const props = {
      ...mockProps,
      products: [...mockProps.products, extraProduct],
    };
    const { container } = render(<ProductList {...props} />);

    const extraProductEmptyCell =
      screen.getAllByRole('rowgroup')[1].childNodes[2].childNodes[1];

    expect(extraProductEmptyCell).toBeEmptyDOMElement();
    expect(screen.getByText('Sample')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
