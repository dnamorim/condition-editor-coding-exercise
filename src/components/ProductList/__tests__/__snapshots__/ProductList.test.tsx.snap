// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList should render an empty cell when a property value doesn't exist 1`] = `
<div>
  <table
    class="ui celled striped table"
  >
    <thead
      class=""
    >
      <tr
        class=""
      >
        <th
          class=""
        >
          Product Name
        </th>
        <th
          class=""
        >
          Color
        </th>
      </tr>
    </thead>
    <tbody
      class=""
    >
      <tr
        class=""
      >
        <td
          class=""
        >
          Foo
        </td>
        <td
          class=""
        >
          Bar
        </td>
      </tr>
      <tr
        class=""
      >
        <td
          class=""
        >
          Bizz
        </td>
        <td
          class=""
        >
          Buzz
        </td>
      </tr>
      <tr
        class=""
      >
        <td
          class=""
        >
          Sample
        </td>
        <td
          class=""
        />
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`ProductList should render correctly 1`] = `
<div>
  <table
    class="ui celled striped table"
  >
    <thead
      class=""
    >
      <tr
        class=""
      >
        <th
          class=""
        >
          Product Name
        </th>
        <th
          class=""
        >
          Color
        </th>
      </tr>
    </thead>
    <tbody
      class=""
    >
      <tr
        class=""
      >
        <td
          class=""
        >
          Foo
        </td>
        <td
          class=""
        >
          Bar
        </td>
      </tr>
      <tr
        class=""
      >
        <td
          class=""
        >
          Bizz
        </td>
        <td
          class=""
        >
          Buzz
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
